src/files/compose/validateDappnodeCompose.ts

```diff
@@ -0,0 +1,237 @@
+  const isCore = getIsCore(manifest);
+
+  // COMPOSE TOP LEVEL restrictions
+
+  validateComposeVersion(composeUnsafe);
+  validateComposeNetworks(composeUnsafe, allowedNetworks);
+
+  // COMPOSE SERVICE LEVEL restrictions
+
+  const cpServicesValues = Object.values(composeUnsafe.services);
+
+  validateComposeServicesKeys(composeUnsafe);
+  validateComposeServicesValues(cpServicesValues, isCore);
+  validateComposeServicesNetworks(
+    cpServicesValues,
+    allowedNetworks,
+    isCore,
+    coreAliases
+  );
+
+  // Compose service volumes: MUST BE CHECK TOGETHER WITH COMPOSE TOP LEVEL VOLUMES
+  const cpServiceVolumes = cpServicesValues
+    .map(composeService => composeService.volumes)
+    .flat();
+
+  if (cpServiceVolumes.length > 0) {
+    validateComposeAndComposeServicesVolumes(
+      composeUnsafe,
+      cpServiceVolumes,
+      isCore
+    );
+  }
+}
+
+/**
+ * Ensures the docker compose version is supported
+ */
+function validateComposeVersion(compose: Compose): void {
+  if (
+    semver.lt(
+      compose.version + ".0",
+      params.MINIMUM_COMPOSE_FILE_VERSION + ".0"
+    )
+  )
+    throw Error(
+      `Compose version ${compose.version} is not supported. Minimum version is ${params.MINIMUM_COMPOSE_FILE_VERSION}`
+    );
+}
+
+/**
+ * Ensures the docker compose networks are whitelisted
+ */
+function validateComposeNetworks(
+  compose: Compose,
+  allowedNetworks: string
+): void {
+  const cpNetworks = compose.networks;
+  if (cpNetworks) {
+    // Check there are only defined whitelisted compose networks
+    if (
+      Object.keys(cpNetworks).some(
+        networkName =>
+          params.DOCKER_WHITELIST_NETWORKS.indexOf(networkName) === -1
+      )
+    )
+      throw Error(`Only docker networks ${allowedNetworks} are allowed`);
+
+    // Check all networks are external
+    if (Object.values(cpNetworks).some(network => network.external === false))
+      throw Error(`Docker internal networks are not allowed`);
+  }
+}
+
+/**
+ * Ensures the compose keys are whitelisted
+ */
+function validateComposeServicesKeys(compose: Compose): void {
+  const composeServicesNames = Object.keys(compose.services);
+
+  for (const composeServiceName of composeServicesNames) {
+    const composeServiceKeys = Object.keys(
+      compose.services[composeServiceName]
+    );
+    if (
+      composeServiceKeys.some(
+        composeServiceKey => params.SAFE_KEYS.indexOf(composeServiceKey) === -1
+      )
+    )
+      throw Error(
+        `Compose service ${composeServiceName} has keys that are not allowed. Allowed keys are: ${params.SAFE_KEYS.join(
+          ","
+        )}`
+      );
+  }
+}
+
+/**
+ * Ensures the compose keys values are valid for dappnode
+ */
+function validateComposeServicesValues(
+  cpServicesValues: ComposeService[],
+  isCore: boolean
+): void {
+  // Check that if defined, the DNS must be the one provided from the bind package
+  if (
+    cpServicesValues.some(
```

Comment:
It would be much easier to check one `ComposeService` at a time instead of calling .some() every time and giving no hint of what specific service is invalid